### ISSUE TITLE
Requirements for Affinity to graduate to Beta and then v1

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#25319]
We'd like to graduate node and pod affinity to Beta in 1.4.

Beta
- [ ] Change annotation key from `scheduler.alpha.kubernetes.io/affinity` to `scheduler.beta.kubernetes.io/affinity`
- [ ] Make performance of pod affinity acceptable (and verify that performance of node affinity is acceptable)
- [ ] Better solution for "first pod problem"
- [ ] Last bullet point here https://github.com/kubernetes/kubernetes/pull/22985#issuecomment-216020283
- [ ] Before allowing users to use it, be sure we won't need to roll back to binary version that doesn't support it

v1
- [ ] Move Affinity to be a field in PodSpec
- [ ] Implement RequiredDuringScheduling
- [ ] Incorporate feedback from users of Beta

Better solution for DoS issues? (or just rely on priority/preemption when we have it?) (see https://github.com/kubernetes/kubernetes/pull/18265#issuecomment-172142482)

cc/ @kevin-wangzefeng @bgrant0607 @wojtek-t 

ref/ #18261 #18265 #24853 #22985 #19758

**Comment by timothysc** [comment]
@rrati is starting on list above for 1.5

**Comment by timothysc** [comment]
@rrati things to keep in mind re(alpha-beta) : #30819 (comment)

**Comment by rrati** [comment]
@davidopp @wojtek-t Do you know the state of the items listed above?  Some of the issues appear to have been either resolved or have had work done on them.

"Before allowing users to use it, be sure we won't need to roll back to binary version that doesn't support it* - Does this have a defined work item?  Is it just a confidence level?

Better solution for "first pod problem" - Has anyone started discussing an alternate plan?

**Comment by davidopp** [comment]
@rrati :

* Change annotation key from `scheduler.alpha.kubernetes.io/affinity` to `scheduler.beta.kubernetes.io/affinity`: There's ongoing discussion about whether to use fields or annotations in beta, in general (if you want me to look up the issue number and kubernetes-dev thread I can, but you probably know what I'm talking about)

* Better solution for "first pod problem": I don't think we really need to do this for Beta, because it is not an API change. If you're wondering what it's referring to, see the section that starts "This last example illustrates" in the [design doc](https://github.com/kubernetes/kubernetes/blob/master/docs/design/podaffinity.md). Today we have implemented the "short term" approach listed there.

* Last bullet point [here](https://github.com/kubernetes/kubernetes/pull/22985#issuecomment-216020283): This is a test improvement. Also not strictly needed for Beta. (Though I suspect we will never do it if we don't do it now.)

* Before allowing users to use it, be sure we won't need to roll back to binary version that doesn't support it: The ongoing discussion related to the first bullet point has also touched on this. Basically it means make sure we do the right thing if someone sets alpha pod affinity in 1.4 and they upgrade to 1.5, sets beta pod affinity in 1.5 and downgrade to 1.4, sets pod affinity in 1.4 and upgrade to 1.5 and then downgrade to 1.4, etc. We have some flexibility in defining "right thing" in the alpha case since we make no guarantees about alpha features, though in the aforementioned discussion Tim was saying we should support the alpha version in a couple of version before we get rid of it, which would mean we need to support both alpha and beta in 1.5.

* Re-enable/fix all tests (see e.g. #26697, #26698): Looks like there's nothing to do here, as I don't see any e2e's still marked Feature:PodAffinity (but can't hurt for you to double-check).

**Comment by ConnorDoyle** [comment]
Made a separate issue for the test improvement: https://github.com/kubernetes/kubernetes/issues/34253

**Comment by timothysc** [comment]
Version moved per sig+community discussion re: alpha-beta transitioning.

**Comment by rrati** [comment]
I would propose converting from annotations to api fields before moving from alpha->beta in a process described in #35518.  Namely, convert from annotations to api fields, strip the annotations implementation out completely, port tests, etc.  Once functionality is confirmed to be at the same level as annotations then promote from alpha->beta.

**Comment by davidopp** [comment]
We should add node and pod affinity to GeneralPredicates when we move them to Beta (this will make Kubelet check them in admission check).

**Comment by davidopp** [comment]
Also, @wojtek-t has suggested we should consider restricting hard anti-affinity to node-level only. The reasons are
* scalability/performance
* for Kubelet admission to check any topology other than node, it would need full cluster state
* it's not clear there are any use cases for hard anti-affinity at topology other than node-level (soft anti-affinity is different; e.g. you might want to spread across zones)

**Comment by wojtek-t** [comment]
>
we should consider restricting hard anti-affinity to node-level only.

We should do it as long as we can (which means we need to make this before doing beta).

**Comment by timothysc** [comment]
Affinity and anti-affinity have now been converted to fields, I'm taking a look at the remaining checklist items.

**Comment by davidopp** [comment]
The only remaining thing I think we need to do is @wojtek-t's suggstion to restrict hard anti-affinity to node TopologyKey only. We already have an [admission controller](
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/antiaffinity/admission.go) that does this by default, but we can't optimize the code with this assumption since the admission controller is optional. I think for Beta we can just modify validation to reject TopologyKey != node -- we can implement the actual optimizations later as those do not change the API.

**Comment by timothysc** [comment]
> it's not clear there are any use cases for hard anti-affinity at topology other than node-level (soft anti-affinity is different; e.g. you might want to spread across zones) 

That's not true at all for some stateful sets and use cases we care about.  We will want hard anti-affinity to spread on various labels (rack, cluster, ...) etc.  It's a key requirement for us.  Seems like a topic for 2day

**Comment by davidopp** [comment]
> That's not true at all for some stateful sets and use cases we care about. We will want hard anti-affinity to spread on various labels (rack, cluster, ...) etc. It's a key requirement for us. Seems like a topic for 2day

Unfortunately I can't remember why @wojtek-t said that restricting to node was better than restricting to a pre-defined set of TopoloyKeys at scheduler startup time (like we do today); I think this is critical information. If it's just about building indexes then the latter is just as good as the former, but I think there's more to it.

**Comment by wojtek-t** [comment]
Computing anti-affinity only for the "node" TopologyKey is relatively simple, as it can be done within pods from a given node only (i.e. it is local computation).
For any other TopologyKey, it is computationally expensive, as in the worst case it takes O(|pods| * |nodes|). And the problem is that it affects all pods, not only those with anti-affinity constraints. So if we want to have anti-affinity in beta/ga on different topology keys, we really need to think how to do it - it cannot be like it currently is.

**Comment by davidopp** [comment]
I think it's fine to go to beta with the restriction TopologyKey == node, and later expand the choices. Today this is enforced using an admission controller that is enabled by default. I think we could just leave that as-is instead of moving the code into validation, if we think we might allow other topology keys later once we can improve the performance.

**Comment by wojtek-t** [comment]
I think today other topology keys are also allowed (though I didn't check it). We for sure allow other topology keys for affinity, and i can't remember any difference in admission code between affinity and anti-affinity. But I'm fine with what @davidopp suggested as a temporary solution.

**Comment by davidopp** [comment]
The admission controller that limits hard pod anti-affinity to node TopologyKey is [here](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/antiaffinity/admission.go).

But you're right, it's not enabled by default. For example none of the cluster/*/config-default.sh (and config-test.sh) scripts list it in the ADMISSION_CONTROL line. So we should either turn this on, or move the check into validation (and then loosen the check if/when we allow other TopologyKey values).

**Comment by davidopp** [comment]
Before this can be considered complete (at Beta) we need to search for and remove all instances of "api.AffinityAnnotationKey", "AffinityAnnotationKey", and "scheduler.alpha.kubernetes.io/affinity" in the codebase.

**Comment by timothysc** [comment]
I'll PR that hopefully tomorrow - https://github.com/kubernetes/kubernetes/issues/40016

**Comment by davidopp** [comment]
Remaining work for moving Affinity to beta:
- [ ] PR to add back support for the alpha annotation under feature-gate and the associated tests (#41215) needs to be completed and merged: assigned to @timothysc 
- [ ] PR to reject empty TopologyKey for the field (but not the annotation), and remove the --failure-domains scheduler command-line flag (#41195) needs to be completed and merged: assigned to @wojtek-t 
- [ ] Make admission controller that limits hard pod anti-afinity to node topology key be enabled by default: assigned to @davidopp
- [ ] Update documentation to use the beta field instead of alpha annotation: currently unassigned

Upgrading to 1.6 will not break existing users of this feature.

**Comment by timothysc** [comment]
@davidopp Should we keep this open or rally on the Feature Repo, as that is where everyone else is tracking.

**Comment by davidopp** [comment]
My understanding is that features repo is for tracking progress not technical details. This issue is kinda a mixture of two, but I worry that if we close it we won't have anywhere to discuss technical details.

**Comment by davidopp** [comment]
ref/ #43203

**Comment by davidopp** [comment]
ref/ #43197

**Comment by davidopp** [comment]
In [this sig-network thread](https://groups.google.com/d/msg/kubernetes-sig-network/4Q1Xk1Jm4-M/bhNrmeiXAgAJ) the issue was raised that we should align the way pod affinity/anti-affinity selects pods (PodAffinityTerm in pkg/api/types.go) with the way NetworkPolicy does (see NetworkPolicyPeer in pkg/apis/extensions/types.go). I linked to the most relevant message in the thread, but the idea is that we'd replace the current "list of namespaces" field with a field of type LabelSelector which would represent a selector over labels on namespaces. 

This would be part of moving affinity to GA, which we should presumably accelerate in an effort to become compatible with what NetworkPolicy is going to do.

I assume the way we'd do it is
1.7 add the new field, and validation enforces that you can set one or the other but not both. mark the "list of namespaces" field as deprecated. People would switch during this release.
1.8 remove the "list of namespaces" field.

cc/ @thockin @kubernetes/sig-scheduling-misc @wojtek-t

**Comment by davidopp** [comment]
BTW one implication of using LabelSelector on namespace labels is that it will be possible to specify "all namespaces" (which we were special-casing when we were doing it as a list of namespaces), because [empty label selector matches all objects](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L762).

In light of this, I don't think it's worth adding "all namespaces" capability to the list-of-namespaces (#43525) as we are going to deprecate list-of-namespaces.

**Comment by wanghaoran1988** [comment]
@davidopp Do we have the final decision here? I saw some different voice in the  sig-network thread about the multiple tenant problem.

**Comment by davidopp** [comment]
I attended the sig-network meeting on April 6 and they decided to go ahead with the label-selector-on-namespace approach.

**Comment by gmarek** [comment]
Before moving PodAntiAffinity to GA (aka v1) we NEED to fix it's performance. It's a non-trivial amount of work, which will probably result in rewriting most of the logic, thus when we do this it probably should bake for one release as a beta feature. @kubernetes/sig-scalability-misc @kubernetes/sig-scheduling-misc

**Comment by timothysc** [comment]
There are no plans to move to GA this cycle that I'm aware of.

**Comment by gyliu513** [comment]
@gmarek does there are any issues related to the performance issues of PodAntiAffinity? I would like to do some investigation if needed.

**Comment by gmarek** [comment]
Yup, I know. I just wanted to leave the note for whoever will be doing it at any point, and for us that we don't forget.

**Comment by geekidentity** [comment]
what time the affinity and anti-affinity release stable version?

**Comment by bsalamat** [comment]
They are currently in Beta. We haven't planned on promoting them to GA because they had many performance/scalability issues. We have addressed some of those issues, but there are still some left to be addressed. No concrete plan yet, but I feel we are getting closer. Maybe we go to GA in 1.14 or 1.15.

**Comment by BarthV** [comment]
> BTW one implication of using LabelSelector on namespace labels is that it will be possible to specify "all namespaces" (which we were special-casing when we were doing it as a list of namespaces), because [empty label selector matches all objects](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L762).
> 
> In light of this, I don't think it's worth adding "all namespaces" capability to the list-of-namespaces (#43525) as we are going to deprecate list-of-namespaces.

Selecting labels over all namespaces would be a really great feature !
As an example, we wanted to avoid some kind of statefulsets running side to side on the same node, but since they are running in their own namespace we're not able to use the podAntiAffinity mecanisms ... So we had to trick this and use state / capacity mecanism which is much less flexible than selectors.

So please keep on mind that it might be a good way to reintroduce cross-namespaced selectors feature in affinity rules :-) , definitely not useful for everyone but a really great tool in some cases !

**Comment by bsalamat** [comment]
@BarthV we heard you all and we will try to add the feature, but we need to be careful here as this feature can increase possibility of DoS attacks. Please see my comment for more info: https://github.com/kubernetes/kubernetes/issues/68827#issuecomment-468470598.

**Comment by pacoxu** [comment]
#97410 need we cleanup those alpha annotations?
